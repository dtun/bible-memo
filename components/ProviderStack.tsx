import { JSXElementConstructor, ReactNode } from "react";

type NoInfer<T> = [T][T extends any ? 0 : 1];

type ContainsChildren = {
  children?: React.ReactNode;
};

function ProviderStack<
  Providers extends [ContainsChildren, ...ContainsChildren[]]
>({
  providers,
  children,
}: {
  providers: {
    [k in keyof Providers]: [
      JSXElementConstructor<Providers[k]>,
      Omit<NoInfer<Providers[k]>, "children">
    ];
  };
  children: ReactNode;
}) {
  let node = children;

  for (let [Provider, props] of providers) {
    node = <Provider {...props}>{node}</Provider>;
  }

  return node;
}

export { ProviderStack };
