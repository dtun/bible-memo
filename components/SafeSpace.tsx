import { View } from "react-native";
import {
  withSafeAreaInsets,
  WithSafeAreaInsetsProps,
} from "react-native-safe-area-context";

function SafeBottom(props: WithSafeAreaInsetsProps) {
  return <View style={{ height: props.insets.bottom }} />;
}

function SafeTop(props: WithSafeAreaInsetsProps) {
  return <View style={{ height: props.insets.top }} />;
}

export const SafeSpace = {
  Top: withSafeAreaInsets(SafeTop),
  Bottom: withSafeAreaInsets(SafeBottom),
};
