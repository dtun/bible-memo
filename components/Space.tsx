import { View } from "./Themed";

function Space({ height }: { height: number }) {
  return (
    <View darkColor="transparent" lightColor="transparent" style={{ height }} />
  );
}

export { Space };
