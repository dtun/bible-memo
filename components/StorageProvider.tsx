import { Platform } from "react-native";
import * as SQLite from "expo-sqlite";
import { Store, createStore } from "tinybase";
import { createLocalPersister } from "tinybase/persisters/persister-browser";
import { Persister } from "tinybase/persisters";
import { createExpoSqlitePersister } from "tinybase/persisters/persister-expo-sqlite";
import {
  Provider,
  useCreatePersister,
  useCreateStore,
} from "tinybase/ui-react";

// Define the shape of your store's data
interface BibleStoreSchema {
  tables: {
    bible: {
      text: string;
      done: boolean;
    };
  };
}

// StorageProvider props
interface StorageProviderProps {
  children: React.ReactNode;
}

// StorageProvider component
const BibleProvider: React.FC<StorageProviderProps> = ({ children }) => {
  const store = useCreateStore<BibleStoreSchema>(() =>
    createStore<BibleStoreSchema>()
  );
  useAndStartPersister(store);

  return <Provider store={store}>{children}</Provider>;
};

const useAndStartPersister = (store: Store<BibleStoreSchema>) =>
  useCreatePersister(
    store,
    (store: Store<BibleStoreSchema>) =>
      Platform.OS === "web"
        ? createLocalPersister(store, "todos")
        : createExpoSqlitePersister(store, SQLite.openDatabaseSync("todos.db")),
    [],
    (persister: Persister<BibleStoreSchema>) =>
      persister.load().then(persister.startAutoSave)
  );

export { BibleProvider };
