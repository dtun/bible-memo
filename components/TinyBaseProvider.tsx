import {
  Provider,
  useCreatePersister,
  useCreateStore,
} from "tinybase/ui-react";
import { createBibleStore, createBiblePersister } from "@/utils/tinybase";

interface TinyBaseProviderProps {
  children: React.ReactNode;
}

export function TinyBaseProvider({ children }: TinyBaseProviderProps) {
  let store = useCreateStore(createBibleStore);

  useCreatePersister(store, createBiblePersister, [], async (persister) => {
    await persister.load();
    await persister.startAutoSave();
  });

  return <Provider store={store}>{children}</Provider>;
}
