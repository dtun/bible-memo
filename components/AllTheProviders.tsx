import {
  DarkTheme,
  DefaultTheme,
  ThemeProvider,
} from "@react-navigation/native";
import { type ReactNode } from "react";

import { useColorScheme } from "./useColorScheme";
import { ProviderStack } from "./ProviderStack";
import { TinyBaseProvider } from "./TinyBaseProvider";

function AllTheProviders({ children }: { children: ReactNode }) {
  let colorScheme = useColorScheme();
  let theme = colorScheme === "dark" ? DarkTheme : DefaultTheme;

  return (
    <ProviderStack
      providers={[
        [ThemeProvider, { value: theme }],
        [TinyBaseProvider, {}],
      ]}
    >
      {children}
    </ProviderStack>
  );
}

export { AllTheProviders };
