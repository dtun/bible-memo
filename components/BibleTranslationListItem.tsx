import { Pressable, StyleSheet } from "react-native";

import { Text, View } from "@/components/Themed";
import { type BibleTranslation } from "@/types";

function BibleTranslationListItem({
  item: bible,
}: {
  item: BibleTranslation & {
    selected?: boolean;
    onPress: (id: string) => void;
  };
}) {
  return (
    <Pressable
      accessibilityLabel={bible.nameLocal}
      accessibilityRole="link"
      accessibilityState={{ selected: bible.selected }}
      hitSlop={24}
      onPress={() => bible.onPress(bible.id)}
      style={({ pressed }) => ({
        opacity: pressed ? 0.5 : 1,
      })}
    >
      <View
        darkColor="transparent"
        lightColor="transparent"
        style={styles.listItem}
      >
        <Text style={styles.title}>{bible.abbreviationLocal}</Text>
        {bible.selected ? <Text style={styles.check}>âœ“</Text> : null}
      </View>
      <Text style={styles.body}>{bible.nameLocal}</Text>
    </Pressable>
  );
}

export { BibleTranslationListItem };

let styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  listEmpty: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    alignSelf: "center",
  },
  body: {
    fontSize: 16,
    marginHorizontal: 4,
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
    marginVertical: 4,
    marginHorizontal: 4,
  },
  check: {
    fontSize: 24,
    fontWeight: "bold",
    marginVertical: 4,
    marginHorizontal: 8,
  },
  separator: {
    marginVertical: 16,
    height: 1,
    width: "100%",
  },
  listItem: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
  },
});
