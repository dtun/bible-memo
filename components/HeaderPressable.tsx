import { forwardRef } from "react";
import { Pressable } from "react-native";

import { Text } from "@/components/Themed";

let HeaderPressable = forwardRef(function HeaderPressable(
  {
    onPress,
    title,
  }: {
    onPress?: () => void;
    title: string;
  },
  _ref
) {
  return (
    <Pressable onPress={onPress}>
      {({ pressed }) => (
        <Text style={{ opacity: pressed ? 0.5 : 1 }}>{title}</Text>
      )}
    </Pressable>
  );
});

export { HeaderPressable };
