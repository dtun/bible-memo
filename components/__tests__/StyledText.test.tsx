import { render, screen } from "@/test-utils";
import { MonoText } from "@/components/StyledText";

it(`renders <MonoText />`, () => {
  render(<MonoText>Snapshot test!</MonoText>);

  expect(screen).toMatchSnapshot();
});

it(`renders <MonoText /> in light`, () => {
  render(<MonoText lightColor="#eee">Snapshot test!</MonoText>);

  expect(screen).toMatchSnapshot();
});

it(`renders <MonoText /> in dark`, () => {
  render(<MonoText darkColor="#000">Snapshot test!</MonoText>);

  expect(screen).toMatchSnapshot();
});
