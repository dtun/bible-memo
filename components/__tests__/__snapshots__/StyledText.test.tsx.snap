// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <MonoText /> 1`] = `
<Text
  style={
    [
      {
        "color": "#000",
      },
      [
        undefined,
        {
          "fontFamily": "SpaceMono",
        },
      ],
    ]
  }
>
  Snapshot test!
</Text>
`;

exports[`renders <MonoText /> in dark 1`] = `
<Text
  style={
    [
      {
        "color": "#000",
      },
      [
        undefined,
        {
          "fontFamily": "SpaceMono",
        },
      ],
    ]
  }
>
  Snapshot test!
</Text>
`;

exports[`renders <MonoText /> in light 1`] = `
<Text
  style={
    [
      {
        "color": "#eee",
      },
      [
        undefined,
        {
          "fontFamily": "SpaceMono",
        },
      ],
    ]
  }
>
  Snapshot test!
</Text>
`;
